### PR TITLE
Retain debug logs for etcd when there is a place to keep them.

### PR DESCRIPTION
**Pull request by mml**
For help debugging #25845

### CODE REVIEW

File: hack/lib/etcd.sh [hunk 1]
```diff
@@ -41,8 +41,13 @@ kube::etcd::start() {
 
   # Start etcd
   ETCD_DIR=$(mktemp -d 2>/dev/null || mktemp -d -t test-etcd.XXXXXX)
-  kube::log::info "etcd -addr ${ETCD_HOST}:${ETCD_PORT} -data-dir ${ETCD_DIR} --bind-addr ${ETCD_HOST}:${ETCD_PORT} >/dev/null 2>/dev/null"
-  etcd -addr ${ETCD_HOST}:${ETCD_PORT} -data-dir ${ETCD_DIR} --bind-addr ${ETCD_HOST}:${ETCD_PORT} >/dev/null 2>/dev/null &
+  if [[ -d "${ARTIFACTS_DIR}" ]]; then
```
Comment (ixdy) [inline hack/lib/etcd.sh]: use default value if `ARTIFACTS_DIR` isn't set - `${ARTIFACTS_DIR:-}`

**Comment by mml** [comment]
@ixdy PTAL

File: hack/lib/etcd.sh [hunk 2]
```diff
@@ -41,8 +41,13 @@ kube::etcd::start() {
...
   ETCD_DIR=$(mktemp -d 2>/dev/null || mktemp -d -t test-etcd.XXXXXX)
-  kube::log::info "etcd -addr ${ETCD_HOST}:${ETCD_PORT} -data-dir ${ETCD_DIR} --bind-addr ${ETCD_HOST}:${ETCD_PORT} >/dev/null 2>/dev/null"
-  etcd -addr ${ETCD_HOST}:${ETCD_PORT} -data-dir ${ETCD_DIR} --bind-addr ${ETCD_HOST}:${ETCD_PORT} >/dev/null 2>/dev/null &
+  if [[ -d "${ARTIFACTS_DIR:-}" ]]; then
+    ETCD_LOGFILE="${ARTIFACTS_DIR}/etcd.$(uname -n).$(id -un).log.DEBUG.$(date +%Y%m%d-%H%M%S).$$"
+  else
+    ETCD_LOGFILE=/dev/null
+  fi
+  kube::log::info "etcd -addr ${ETCD_HOST}:${ETCD_PORT} -data-dir ${ETCD_DIR} --bind-addr ${ETCD_HOST}:${ETCD_PORT} --debug > \"${ETCD_LOGFILE}\" 2>/dev/null"
+  etcd -addr ${ETCD_HOST}:${ETCD_PORT} -data-dir ${ETCD_DIR} --bind-addr ${ETCD_HOST}:${ETCD_PORT} --debug > "${ETCD_LOGFILE}" 2>/dev/null &
```
Comment (ixdy) [inline hack/lib/etcd.sh]: it looks like debug output is printed to stderr, possibly?

**Comment by mml** [comment]
This didn't get flagged as a failure, and it's clearly not due to my PR, but...

```
 425-          <*errors.errorString | 0xc820b5b9f0>: {
 426-              s: "expected metrics for kubelet",
 427-          }
 428-      to be nil
 429-
 430-      /var/lib/jenkins/workspace/node-pull-build-e2e-test@3/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:162
 431-------------------------------
 432-SSS••
 433-
 434:Summarizing 1 Failure:
 435-I0607 00:41:54.710908     924 e2e_node_suite_test.go:112] Tests Finished
 --
 512-          <*errors.errorString | 0xc820b5b9f0>: {
 513-              s: "expected metrics for kubelet",
 514-          }
 515-      to be nil
 516-
 517-      /var/lib/jenkins/workspace/node-pull-build-e2e-test@3/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:162
 518-------------------------------
 519-SSS••
 520-
 521:Summarizing 1 Failure:
 522-I0607 00:41:54.710908     924 e2e_node_suite_test.go:112] Tests Finished
 --
 525-/var/lib/jenkins/workspace/node-pull-build-e2e-test@3/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/kubelet_test.go:162
 526-
 527-Ran 9 of 16 Specs in 418.217 seconds
 528-FAIL! -- 8 Passed | 1 Failed | 0 Pending | 7 Skipped --- FAIL: TestE2eNode (418.22s)
 529-FAIL
 530-
 531-================================================================
 532-[PostBuildScript] - Execution post build scripts.
 533-[node-pull-build-e2e-test@3] $ /bin/bash -xe /tmp/hudson8568034669811214900.sh
 534:+ [[ -x ./hack/jenkins/upload-to-gcs.sh ]]
```

**Comment by ixdy** [comment]
Yeah, that's a KI we're working on fixing.

And it looks like stdout is not what you want - the uploaded etcd debug logs are all empty.

**Comment by mml** [comment]
> And it looks like stdout is not what you want - the uploaded etcd debug logs are all empty.

That is weird.  I could swear this did exactly what I expected on my workstation.  Let me check again.

**Comment by mml** [comment]
@ixdy not sure what happened before, but yes, changed to stderr.

**Comment by ixdy** [comment]
LGTM

**Pull request merged**